### PR TITLE
Add vaccination models.

### LINKED ISSUE #69: Add models for vaccination.
### Describe the solution you'd like

- models for vaccinations are like events, but instead of spreading infections, shots are distributed.
- Add documentation which explains
  - how to distribute vaccinations
  - apply a limit on the number of available shots
  - how to apply a second shot
  - compare one scenario with the regular second shot and one scenario where the second shot is delayed.

### PR DESCRIPTION
**Pull request by tobiasraabe**
Closes #69.

The `simulate` function receives a new argument called `vaccination_model` which is a single function returning a boolean indicator of individuals who get vaccinated.

This starts a countdown at which end an individual is immune. The countdown can be set in the parameters with `("cd_is_immune_by_vaccine", "all", -1)`.

The new columns are:

- `ever_vaccinated` indicates whether individuals recevied vaccines.
- `newly_vaccinated` indicates individuals who received a vaccine in this period
- `cd_is_immune_by_vaccine` countdown which if expired sets `immune` to `True`.
- `cd_is_immune_by_vaccine_draws` holds the duration until vaccines kick in.

### CODE REVIEW

**Review by janosg — COMMENTED** [review]
- Instead of the `is_vaccinated` and `received_vaccine` columns I would propose `ever_vaccinated` and `newly_vaccinated` as counterparts to `ever_infected` and `newly_infected`. In contrast to `is_vaccinated`, `ever_vaccinated` becomes True the moment one receives a vaccine. That way it becomes also easier not to vaccinate people twice (in the phase where they are already vaccinated but not yet immune).
- For initial infections and immunity we pass in DataFrames. Here we use a function that basically returns the DataFrame column by column. I first did not like it, but now I think it's better because it makes it easier only to vaccinate people who are not in quarantine (or more generally to react to health states). We should think about changing the interface for initial infections in the same direction. But not in this PR, of course.

**Comment by tobiasraabe** [comment]
Hi @janosg,

I added/renamed the variables.

Regarding the function in the initial conditions, I felt it is easier and more similar to testing to have it periodically. On the other hand, if you want to start with pre-existing vaccinations and have a shorter burn-in period, you might have no vaccinated people at the start of the simulation.

File: src/sid/update_states.py [hunk 1]
```diff
@@ -158,3 +164,16 @@ def _update_info_on_new_tests(
...
 
     return states
+
+
+def _update_info_on_new_vaccinations(
+    states: pd.DataFrame, newly_vaccinated: pd.Series
+) -> pd.DataFrame:
+    """Activate the counter for immunity by vaccinations."""
+    states["newly_vaccinated"] = newly_vaccinated
+    states.loc[newly_vaccinated, "ever_vaccinated"] = newly_vaccinated
```
Comment (janosg) [inline src/sid/update_states.py:174]: Would `states.loc[newly_vaccinated, "ever_vaccinated"] = True` also work? If so, I find it more readable. I am not even sure why your version runs. If `newly_vaccinated` has False entries, I would expect a length mismatch.

**Review by janosg — APPROVED** [review]
One more comment. Rest is very nice!

File: src/sid/update_states.py (hunk 1 above)
Comment (tobiasraabe) [reply src/sid/update_states.py:174]: I think pandas would match via indices, but I prefer your syntax as well.

Comment (tobiasraabe) [reply src/sid/update_states.py]: Actually, it is not a guess. It is tested that it works like this. And its something I appreciated when using other languages.

Comment (janosg) [reply src/sid/update_states.py]: Makes sense. Thanks!

**Pull request merged**